File: sol/solution/0001-0100/0035. Search Insert Position/Solution.ex
defmodule Solution do
  @spec search_insert(nums :: [integer], target :: integer) :: integer
  def search_insert(nums, target) do
    search(nums, target, 0, length(nums) - 1)
  end
  defp search(nums, target, left, right) when left <= right do
    mid = div(left + right, 2)
    val = Enum.at(nums, mid)
    cond do
      val == target -> mid
      val < target -> search(nums, target, mid + 1, right)
      true -> search(nums, target, left, mid - 1)
    end
  end
  defp search(_nums, _target, left, _right), do: left
end

File: sol/solution/0001-0100/0035. Search Insert Position/Solution.erl
-spec search_insert([integer()], integer()) -> integer().
search_insert(Nums, Target) ->
    search(Nums, Target, 0, length(Nums) - 1).
search(Nums, Target, Left, Right) when Left =< Right ->
    Mid = (Left + Right) div 2,
    Val = lists:nth(Mid + 1, Nums),
    case Val of
        Target -> Mid;
        _ when Val < Target -> search(Nums, Target, Mid + 1, Right);
        _ -> search(Nums, Target, Left, Mid - 1)
    end;
search(_, _, Left, _) ->
    Left.